06.Lesson_ThinkingProperties/ThinkingPropertiesExercise/MeetingScheduler/meetingscheduler.spec

```diff
@@ -0,0 +1,210 @@
+    env e1; calldataarg args1;
+    env e2; calldataarg args2;
+    method f1;
+    method f2;
+
+    // call the parametric function on two different meetings
+    require meetingId1 != meetingId2;
+
+    // get the state of the 2 meetings, #1 and #2, before making a function call
+    
+    uint256 startTime1Before; uint256 endTime1Before;
+    startTime1Before = getStartTimeById(meetingId1);
+    endTime1Before = getEndTimeById(meetingId1);
+
+    uint8 status2Before; uint256 startTime2Before; uint256 endTime2Before; uint256 numParticipants2Before;
+    status2Before = getStateById(meetingId2);
+    startTime2Before = getStartTimeById(meetingId2);
+    endTime2Before = getEndTimeById(meetingId2);
+    numParticipants2Before = getNumOfParticipents(meetingId2);
+
+    // make the state change on #1
+    callFunctionWithParams(f1, meetingId1, startTime1Before, endTime1Before);
+
+    // check that calls to #1 had no effect on #2
+    uint8 status2After; uint256 startTime2After; uint256 endTime2After; uint256 numParticipants2After;
+    status2After = getStateById(meetingId2);
+    startTime2After = getStartTimeById(meetingId2);
+    endTime2After = getEndTimeById(meetingId2);
+    numParticipants2After = getNumOfParticipents(meetingId2);
+    assert (status2Before == status2After, "status changed");
+    assert (startTime2Before == startTime2After, "startTime changed");
+    assert (endTime2Before == endTime2After, "endTime changed");
+    assert (numParticipants2Before == numParticipants2After, "numParticipants changed");
+}
+
+
+// TODO: i don't know how to compare reachability of 2 function calls (1 by the organizer, the other
+// by the non-organizer using the `storage` syntax. Gotta ask Mike)
+// all function calls that succeed for a non-organizer would all succeed for an organizer
+// rule nonOrganizerCallsAlwaysSucceedForOrganizer(address nonOrganizer, address organizer) {
+//     env e; calldataarg args;
+
+//     require e.msg.sender == organizer;
+
+
+//     // I wanna check, if we start in the same state, and then consider 2 possibilities:
+//     // 1. an organizer calls a function
+//     // 2. a non-organizer calls a function
+//     // then there should be no instances where non-organizer succeeds, and organizer fails
+//     // another way of saying this is that the set of successful function calls for organizer
+//     // is a superset of the non-organizer's set of successful function calls
+
+
+//     // take a snapshot of the state so we can compare 
+//     storage initial = lastStorage;
+// }
+
+// after a transition to the ENDED (status = 3), then block.timestamp > endTime
+rule stateOfEnded(uint256 meetingId) {
+    env e;
+
+    uint256 stateBefore = getStartTimeById(meetingId);
+    // we want to ensure we're in the right initial state, which if
+    // we're trying to transition to ENDED means we must be in the STARTED
+    // (status = 2) state
+    require stateBefore == 2;
```

Comment:
I would not add any requires here.  Instead, I would just do what the english text says: call a function, and check that if there was a transition and the state after is ENDED then the timestamp is appropriate.

So, instead of requiring `stateBefore == 2`, I would just require `stateBefore != ENDED`.

Reply:
Note that you can access enums defined in the contracts, see https://docs.certora.com/en/latest/docs/cvl/types.html#user-defined-types